File: agent-runtime/src/protocols/erc8004/validation.ts
/**
 * ERC-8004 Validation Management
 *
 * Validator hook integration for attestation and verification.
 */

import { ethers } from 'ethers';
import { type ChainConfig, resolveChain } from '../../utils/chains.js';
import {
  validationRegistry,
  createSigner,
} from '../../utils/contracts.js';
import type { ValidationRecord, ValidationStatus } from './types.js';

export class ValidationManager {
  private readonly chain: ChainConfig;
  private readonly privateKey?: string;

  constructor(chain: string | number, privateKey?: string) {
    this.chain = resolveChain(chain);
    this.privateKey = privateKey;
  }

  /**
   * Check if an agent is validated for a given attestation type.
   */
  async isValidated(
    agentId: number,
    attestationType: string
  ): Promise<boolean> {
    if (!this.chain.contracts.validation) {
      throw new Error(`Validation registry not deployed on ${this.chain.name}`);
    }
    const registry = validationRegistry(this.chain);
    return registry.isValidated(agentId, attestationType);
  }

  /**
   * Get validation details for an agent.
   */
  async getValidation(
    agentId: number,
    attestationType: string
  ): Promise<ValidationRecord | null> {
    if (!this.chain.contracts.validation) {
      throw new Error(`Validation registry not deployed on ${this.chain.name}`);
    }
    const registry = validationRegistry(this.chain);
    try {
      const [validator, attestationData, timestamp] =
        await registry.getValidation(agentId, attestationType);
      return {
        validator,
        attestationType,
        attestationData: ethers.hexlify(attestationData),
        timestamp: Number(timestamp),
      };
    } catch {
      return null;
    }
  }

  /**
   * Submit a validation attestation for an agent.
   * Note: Only authorized validators can call this.
   */
  async validate(
    agentId: number,
    attestationType: string,
    attestationData: string
  ): Promise<string> {
    if (!this.privateKey) {
      throw new Error('Private key required to submit validation');
    }
    if (!this.chain.contracts.validation) {
      throw new Error(`Validation registry not deployed on ${this.chain.name}`);
    }

    const signer = createSigner(this.privateKey, this.chain);
    const registry = validationRegistry(this.chain, signer);

    const tx = await registry.validate(
      agentId,
      attestationType,
      ethers.toUtf8Bytes(attestationData)
    );
    const receipt = await tx.wait();
    return receipt.hash;
  }

  /**
   * Get full validation status for an agent across common attestation types.
   */
  async getStatus(agentId: number): Promise<ValidationStatus> {
    const commonTypes = ['identity', 'capability', 'security', 'compliance'];
    const records: ValidationRecord[] = [];

    for (const type of commonTypes) {
      try {
        const isValid = await this.isValidated(agentId, type);
        if (isValid) {
          const record = await this.getValidation(agentId, type);
          if (record) records.push(record);
        }
      } catch {
        // Skip attestation types that fail
      }
    }

    return {
      agentId,
      validated: records.length > 0,
      records,
    };
  }
}

File: erc8004-agents/sdks/python/erc8004/validation.py
"""ValidationRegistry interactions."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from web3 import AsyncWeb3
from web3.contract import AsyncContract

from erc8004.contracts import VALIDATION_ABI

if TYPE_CHECKING:
    from erc8004.types import ChainConfig


@dataclass
class ValidationRecord:
    """A validation record from the registry."""

    validation_id: int
    agent_id: int
    validation_type: str
    validator: str
    timestamp: int


class ValidationRegistry:
    """Interact with the ERC-8004 ValidationRegistry contract."""

    def __init__(self, w3: AsyncWeb3, chain: ChainConfig) -> None:
        if not chain.validation_registry:
            raise ValueError(f"No ValidationRegistry deployed on {chain.name}")
        self._w3 = w3
        self._chain = chain
        self._contract: AsyncContract = w3.eth.contract(
            address=w3.to_checksum_address(chain.validation_registry),
            abi=VALIDATION_ABI,
        )

    async def validate(
        self,
        agent_id: int,
        validation_type: str,
        evidence: bytes = b"",
    ) -> int:
        """Submit a validation for an agent.

        Args:
            agent_id: The agent token ID.
            validation_type: Type of validation (e.g., 'identity', 'capability').
            evidence: Validation evidence as bytes.

        Returns:
            The validation record ID.
        """
        tx = await self._contract.functions.validate(  # type: ignore[attr-defined]
            agent_id, validation_type, evidence
        ).transact()
        receipt = await self._w3.eth.wait_for_transaction_receipt(tx)

        # Parse Validated event
        try:
            logs = self._contract.events.Validated().process_receipt(receipt)  # type: ignore[attr-defined]
            if logs:
                return int(logs[0]["args"]["validationId"])
        except Exception:
            pass

        raise RuntimeError("Could not parse validation ID from receipt")

    async def is_valid(self, agent_id: int, validation_type: str) -> bool:
        """Check if an agent has a valid validation of a given type.

        Args:
            agent_id: The agent token ID.
            validation_type: Type of validation to check.

        Returns:
            True if the agent is validated.
        """
        return await self._contract.functions.isValid(  # type: ignore[attr-defined]
            agent_id, validation_type
        ).call()

    async def get_validation(self, validation_id: int) -> ValidationRecord:
        """Get details of a specific validation record.

        Args:
            validation_id: The validation record ID.

        Returns:
            ValidationRecord with full details.
        """
        result = await self._contract.functions.getValidation(validation_id).call()  # type: ignore[attr-defined]
        return ValidationRecord(
            validation_id=validation_id,
            agent_id=result[0],
            validation_type=result[1],
            validator=result[2],
            timestamp=result[3],
        )

    async def get_version(self) -> str:
        """Get the contract version string."""
        return await self._contract.functions.getVersion().call()  # type: ignore[attr-defined]
